src/main/java/net/neoforged/neoforge/common/crafting/CraftingHelper.java

```diff
@@ -20,7 +20,14 @@ public class CraftingHelper {
     public static Codec<Ingredient> makeIngredientCodec(boolean allowEmpty) {
         var compoundIngredientCodec = Codec.lazyInitialized(() -> allowEmpty ? CompoundIngredient.DIRECT_CODEC : CompoundIngredient.DIRECT_CODEC_NONEMPTY);
         return Codec.either(compoundIngredientCodec, makeIngredientMapCodec().codec())
-                .xmap(either -> either.map(ICustomIngredient::toVanilla, i -> i), ingredient -> {
+                .xmap(either -> either.map(custom -> {
+                    // Convert empty compound ingredients back to Ingredient.EMPTY, for the isEmpty() check to return true.
+                    if (custom instanceof CompoundIngredient compound && compound.children().isEmpty()) {
+                        return Ingredient.EMPTY;
+                    } else {
+                        return custom.toVanilla();
+                    }
```

Comment:
Given `toVanila` is an overridable method, why not just override it on `CompoundIngredient`?

Reply:
This comment made me realize that `CompoundIngredient` should never be empty. I have adjusted the code accordingly.